### ISSUE TITLE
Tensor Parallel + FSDP

### ISSUE THREAD

**Issue by EvgenijGod** [pytorch/pytorch#117843]
## Issue description

I am trying to train large language models using FSDP and tensor parallel technology.

I have explored the [native capabilities](https://pytorch.org/docs/stable/distributed.tensor.parallel.html#tensor-parallelism-torch-distributed-tensor-parallel) of torch, I mean `parallelize_module` and `ColwiseParallel`, `RowwiseParallel`. Apparently, these methods are good when the model fits completely into the GPU memory, but they do not imply other initialization options and have problems with Attention accroding to documentation (maybe I'm wrong, but anyway).

My solution was to use tensor parallel implementation from [Nvidia Megatron code](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/layers.py). And using 2-D `DeviceMesh` (one dimension for data parallel group, second dimension for tensor parallel group), I could manage the data & tensor parallel groups of the GPU`s. Sharding of the part of the model corresponding to each data parallel group is performed using the process_group parameter in FullyShardedDataParallel. Thus, FSDP does not know about tensor parallel, but when executing forward and backward tensor parallel FSDP modules communicate with each other in the way it is implemented in the megatron implementation inside Parallel layers.

However, judging by the saved checkpoints, an error occurs somewhere and the gradient values for duplicate layers differ for different FSDP modules corresponding to data parallel parts of the model (in order inputs and loss functions are same here).

For simplicity, I started an experiment for a model with two layers - `Embedding layer` + [ColumnParallel](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/layers.py#L543) layer from Megatron with tensor parallel mode = 2. Column Parallel gathers the output, so layer`s output is exactly equal of applying 2 times larger nn.Linear layer. For two FSDP modules - the Embedding layer shape and initalization are equal and loss function values are equal, but gradient for Embedding layer and values diverges after 1000 iterations  

## Code example

```
...
config.tensor_parallel_size = 2
...
class MyModel(...):
    def __init__(self, config):

        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, config.pad_token_id)
        self.lm_head = ColumnParallelLinear(
                config.hidden_size,
                config.vocab_size,
                config=config,
                bias=False,
                gather_output=True,
                skip_bias_add=True,
            )

    def forward(input_ids, labels):
        hidden_stated = self.embed_tokens(input_ids)
        logits = self.lm_head(hidden_stated)
        ...
        return CrossEntropyLoss()(shift_logits, shift_labels)

model = MyModel()
device_mesh = DeviceMesh("cuda", mesh=torch.arange(torch.dist.get_world_size()).reshape(-1, config.tensor_parallel_size))
_DATA_MODEL_PARALLEL_GROUP, _TENSOR_MODEL_PARALLEL_GROUP = device_mesh.get_dim_groups()
...
model = FullyShardedDataParallel(
    ...,
    process_group=_DATA_MODEL_PARALLEL_GROUP,
    ...,
)

# Training loop
...
input = get_input_for_tensor_parallel_group()
loss = model(input)
loss.backward()
....
```

I apologize that I cannot provide a full year of study, but it is equivalent to what is used in MosaicML`s llm-foundry & composer [here](https://github.com/mosaicml/composer/blob/dev/composer/trainer/trainer.py#L2050)

For 8 GPUs data parallel groups are [0,2,4,6], [1,3,5,7] and tensor parallel groups are [0,1], [2,3], [4,5], [6,7]. So, function `get_input_for_tensor_parallel_group` provides the same input for all ranks in tensor parallel group, so the loss function value is same. But gradients diverge after 1000 steps 

What can be wrong in my FSDP + tensor parallel scheme?

## System Info
Result of `collect_env.py`
https://gist.github.com/EvgenijGod/b0b0a0be8948a9f2f6ba18dee7558c80
